lib/xapi-stdext-date/date.ml

```diff
@@ -38,35 +28,86 @@ let rfc822_of_float x =
 
 let rfc822_to_string x = x
 
+(* ==== ISO8601/RFC3339 ==== *)
+type old_time = string
+type error = string
+type p_time = Ptime.t * Ptime.tz_offset_s option
+
+(** the name doesn't make much sense anymore, but is kept for compatibility reasons
+  * if string is received in rfc3339 format, timezone will be preserved between to/from string
+    but not once converted to a float
+*)
+type iso8601 =
+  | P_time of p_time        (* rfc3339 - accepts only the date in YYYY-MM-DD form *)
+  | Old of old_time * error (* iso8601 - accepts both YYYYMMDD & YYYY-MM-DD *)
+
+ let of_string x =
+  let r = Ptime.of_rfc3339 x in
+  match r with
+  | Ok (t, tz, _) -> P_time (t, tz) (** prefer to use ptime
+                                      * it will only accept dashes *)
+  | Error e -> match Ptime.rfc3339_error_to_msg r with
+               | Error (`Msg err_msg) -> Old (x, err_msg)
+               | Ok x -> failwith "impossible" (* ptime providing bad api *)
```

Comment:
I'd suggest to use `"impossible: "^__LOC__` - so we don't depend on the stack trace.

Reply:
i realised that you can simply the code to avoid this